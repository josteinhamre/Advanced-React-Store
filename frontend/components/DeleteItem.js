import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import { ALL_ITEMS_QUERY } from './Items';

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

class DeleteItem extends Component {
  update = (cache, payload) => {
    // Manualy update the cache on the client, so it matches the server.
    // Read the cache for the items we want
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    // Filter out the deleted item.
    data.items = data.items.filter(
      item => item.id !== payload.data.deleteItem.id
    );
    // Write it back to the cache
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
  };

  render() {
    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
      >
        {(deleteItem, { error }) => (
          <button
            type="button"
            onClick={() => {
              if (
                window.confirm('Are you sure you want to delete this item?')
              ) {
                deleteItem().catch(err => {
                  alert(err.message);
                });
              }
            }}
          >
            {this.props.children}
          </button>
        )}
      </Mutation>
    );
  }
}

export default DeleteItem;
export { DELETE_ITEM_MUTATION };
