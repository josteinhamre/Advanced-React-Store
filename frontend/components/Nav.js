import Link from 'next/link'
import NavStyles from '../components/styles/NavStyles'
import User from './User'

const Nav = () => (
    <NavStyles>
        <User>
            {({data: { me }}) => {
                console.log(me)
                if (me) return <p>{me.name}</p>;
                return null;
            }}
        </User>
        <Link href="/items">
            <a>Shop</a>
        </Link>
        <Link href="/sell">
            <a>Sell</a>
        </Link>
        <Link href="/signup">
            <a>Signup</a>
        </Link>
        <Link href="/ordes">
            <a>Orders</a>
        </Link>
        <Link href="/me">
            <a>Account</a>
        </Link>
    </NavStyles>
);

export default Nav;
