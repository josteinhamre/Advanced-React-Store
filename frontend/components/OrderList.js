import React from 'react';
import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import { formatDistance } from 'date-fns';
import Link from 'next/link';
import styled from 'styled-components';
import formatMoney from '../lib/formatMoney';
import Error from './ErrorMessage';
import OrderItemStyles from './styles/OrderItemStyles';

const USER_ORDERS_QUERY = gql`
  query USER_ORDERS_QUERY {
    orders(orderBy: createdAt_DESC) {
      id
      total
      createdAt
      items {
        id
        title
        price
        description
        quantity
        image
      }
    }
  }
`;

const OrderUl = styled.ul`
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
  grid-gap: 4rem px;
`;

class OrderList extends React.Component {
  render() {
    return (
      <Query query={USER_ORDERS_QUERY}>
        {({ data: { orders }, error, loading }) => {
          if (loading) return <p>Loading...</p>;
          if (error) return <Error error={error.message} />;
          return (
            <div>
              <h2>
                You have {orders.length} order{orders.length > 1 ? 's' : ''}
              </h2>
              <OrderUl>
                {orders.map(order => (
                  <OrderItemStyles key={order.id}>
                    <Link
                      href={{
                        pathname: '/order',
                        query: {
                          id: order.id,
                        },
                      }}
                    >
                      <a>
                        <div className="order-meta">
                          <p>
                            {order.items.reduce(
                              (sum, item) => sum + item.quantity,
                              0
                            )}{' '}
                            items
                          </p>
                          <p>{order.items.length} Products</p>
                          <p>{formatDistance(order.createdAt, new Date())}</p>
                          <p>{formatMoney(order.total)}</p>
                        </div>
                        <div className="images">
                          {order.items.map(item => (
                            <img
                              key={item.id}
                              src={item.image}
                              alt={item.title}
                            />
                          ))}
                        </div>
                      </a>
                    </Link>
                  </OrderItemStyles>
                ))}
              </OrderUl>
            </div>
          );
        }}
      </Query>
    );
  }
}

export default OrderList;
export { USER_ORDERS_QUERY };
