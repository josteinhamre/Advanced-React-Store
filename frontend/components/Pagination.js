import React from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import Head from 'next/head';
import Link from 'next/link';
import { perPage } from '../config';
import Error from './ErrorMessage';
import PaginationStyles from './styles/PaginationStyles';

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    itemsConnection {
      aggregate {
        count
      }
    }
  }
`;

const Pagination = props => (
  <Query query={PAGINATION_QUERY}>
    {({ data, loading, error }) => {
      if (loading) return <p>I am still looking for more data!</p>;
      if (error) return <Error error={error} />;
      const { count } = data.itemsConnection.aggregate;
      const pages = Math.ceil(count / perPage);
      const { page } = props;
      return (
        <PaginationStyles>
          <Head>
            <title>
              Sick Fits! Page {page} of {pages}
            </title>
          </Head>
          <Link
            prefetch
            href={{
              pathname: 'items',
              query: { page: page - 1 },
            }}
          >
            <a className="prev" aria-disabled={page <= 1}>
              ←
            </a>
          </Link>
          <p>
            Page {page} of {pages}
          </p>
          <p>{count} items total</p>
          <Link
            prefetch
            href={{
              pathname: 'items',
              query: { page: page + 1 },
            }}
          >
            <a className="prev" aria-disabled={page >= pages}>
              →
            </a>
          </Link>
        </PaginationStyles>
      );
    }}
  </Query>
);

export default Pagination;
