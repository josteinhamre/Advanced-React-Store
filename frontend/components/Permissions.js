import { Query } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import Error from './ErrorMessage';
import Table from './styles/Table';
import SickButton from './styles/SickButton';

const possiblePermissions = [
  'ADMIN',
  'USER',
  'ITEMCREATE',
  'ITEMUPDATE',
  'ITEMDELETE',
  'PERMISSIONUPDATE',
];

const ALL_USERS_QUERY = gql`
  query ALL_USERS_QUERY {
    users {
      id
      email
      name
      permissions
    }
  }
`;

const Permissions = props => (
  <Query query={ALL_USERS_QUERY}>
    {({ data, loading, error }) => (
      <div>
        <Error error={error} />
        <div>
          <h2>Manage Permissions</h2>
          <Table>
            <thead>
              <tr>
                <th>Name</th>
                <th>Email</th>
                {possiblePermissions.map(permission => (
                  <th key={permission}>{permission}</th>
                ))}
                <th>Confirm</th>
              </tr>
            </thead>
            <tbody>
              {data.users.map(user => (
                <UserPermissions key={user.id} user={user} />
              ))}
            </tbody>
          </Table>
        </div>
      </div>
    )}
  </Query>
);

class UserPermissions extends React.Component {
  static propTypes = {
    user: PropTypes.shape({
      name: PropTypes.string,
      email: PropTypes.string,
      id: PropTypes.string,
      permission: PropTypes.array,
    }).isRequired,
  };

  state = {
    permissions: this.props.user.permissions,
  };

  handlePermissionChange = e => {
    console.log(this.state.permissions);
    const checkbox = e.target;
    let updatedPermissions = [...this.state.permissions];
    console.log(checkbox);
    console.log(updatedPermissions);
    if (checkbox.checked) {
      updatedPermissions.push(checkbox.value);
    } else {
      updatedPermissions = updatedPermissions.filter(
        permission => permission !== checkbox.value
      );
    }
    console.log(updatedPermissions);
    this.setState({ permissions: updatedPermissions });
    console.log(this.state.permissions);
  };

  render() {
    const { user } = this.props;
    return (
      <tr>
        <td>{user.name}</td>
        <td>{user.email}</td>
        {possiblePermissions.map(permission => (
          <td key={`${user.id}-${permission}`}>
            <label htmlFor={`${user.id}-permission-${permission}`}></label>
            <input
              type="checkbox"
              checked={this.state.permissions.includes(permission) && true}
              value={permission}
              onChange={this.handlePermissionChange}
            ></input>
          </td>
        ))}
        <td>
          <SickButton>Update</SickButton>
        </td>
      </tr>
    );
  }
}

export default Permissions;
