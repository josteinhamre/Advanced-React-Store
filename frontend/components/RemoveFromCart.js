import { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { CURRENT_USER_QUERY } from './User';

const REMOVE_FROM_CART_MUTATION = gql`
  mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`;

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${props => props.theme.red};
    cursor: pointer;
  }
`;

class RemoveFromCart extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
  };

  update = (cache, payload) => {
    // Manualy update the cache on the client, so it matches the server.
    // Read the cache for the items we want
    const data = cache.readQuery({ query: CURRENT_USER_QUERY });
    // Filter out the deleted item.
    data.me.cart = data.me.cart.filter(
      item => item.id !== payload.data.removeFromCart.id
    );
    // Write it back to the cache
    cache.writeQuery({ query: CURRENT_USER_QUERY, data });
  };

  render() {
    return (
      <Mutation
        mutation={REMOVE_FROM_CART_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
        optimisticResponse={{
          __typename: 'Mutation',
          removeFromCart: {
            __typename: 'CartItem',
            id: this.props.id,
          },
        }}
      >
        {(removeFromCart, { loading, error }) => (
          <BigButton
            title="Delete Item"
            type="button"
            disabled={loading}
            onClick={() => {
              removeFromCart().catch(err => alert(err.message));
            }}
          >
            &times;
          </BigButton>
        )}
      </Mutation>
    );
  }
}

export default RemoveFromCart;
export { REMOVE_FROM_CART_MUTATION };
