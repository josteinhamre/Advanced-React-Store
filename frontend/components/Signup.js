import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Form from './styles/Form';
import Error from './ErrorMessage';

const SIGNUP_MUTATION = gql`
    mutation SIGNUP_MUTATION(
        $name: String!
        $email: String!
        $password: String!
    ) {
        signUp(
            name: $name
            email: $email
            password: $password
        ) {
            id
            email
            name
        }
    }
`;

class Signup extends Component {
    state = {
        name: '',
        email: '',
        password: ''
    }
    saveToState = (e) => {
        this.setState({ [e.target.name]: e.target.value });
    }
    render() {
        return (
            <Mutation 
                mutation={SIGNUP_MUTATION}
                variables={this.state}
            >
                {(signup, { error, loading }) => {
                    return (
                        <Form method="post" onSubmit={async e => {
                            e.preventDefault();
                            await signup();
                            this.setState({ name: '', email: '', password: ''});
                        }}>
                            <fieldset disabled={loading} aria-busy={loading}>
                                <h2>Sign up for an account</h2>
                                <Error error={error} />
                                <label htmlFor="email">
                                    Email
                                    <input 
                                        type="email" 
                                        id="email"
                                        name="email"
                                        placeholder="Email" 
                                        required 
                                        value={this.state.email}
                                        onChange={this.saveToState}    
                                        />

                                </label>
                                <label htmlFor="name">
                                    Name
                                    <input 
                                        type="text" 
                                        id="name"
                                        name="name"
                                        placeholder="Name" 
                                        required 
                                        value={this.state.name}
                                        onChange={this.saveToState}    
                                        />

                                </label>
                                <label htmlFor="password">
                                    Password
                                    <input 
                                        type="password" 
                                        id="password"
                                        name="password"
                                        placeholder="Password" 
                                        required 
                                        value={this.state.password}
                                        onChange={this.saveToState}    
                                        />

                                </label>
                                <button type="submit">submit</button>
                            </fieldset>
                        </Form>
                        )
                }}
            </Mutation>
        )
    }
}

export default Signup;