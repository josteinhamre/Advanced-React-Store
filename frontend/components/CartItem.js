import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import formatMoney from '../lib/formatMoney';
import RemoveFromCart from './RemoveFromCart';

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightGrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;

  img {
    object-fit: cover;
    margin-right: 1rem;
  }
  h3,
  p {
    margin: 0;
  }
`;

const CartItem = ({ cartItem }) => (
  <CartItemStyles>
    <img width="100" src={cartItem.item.image} alt={cartItem.item.title} />
    <div className="cart-item-details">
      <h3>{cartItem.item.title}</h3>
      <p>
        {formatMoney(cartItem.item.price * cartItem.quantity)}
        {' - '}
        <em>
          {cartItem.quantity} &times;
          {formatMoney(cartItem.item.price)} each
        </em>
      </p>
    </div>
    <div>
      <RemoveFromCart id={cartItem.id} />
    </div>
  </CartItemStyles>
);

CartItem.prototype = {
  cartItem: PropTypes.object.isRequired,
};

export default CartItem;
