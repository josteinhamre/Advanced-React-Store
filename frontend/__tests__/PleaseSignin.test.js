import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import { CURRENT_USER_QUERY } from '../components/User';
import PleaseSignInComponent from '../components/PleaseSignIn';
import { fakeUser } from '../lib/testUtils';

const notSignedInMocks = [
  {
    request: {
      query: CURRENT_USER_QUERY,
    },
    result: {
      data: {
        me: null,
      },
    },
  },
];

const signedInMocks = [
  {
    request: {
      query: CURRENT_USER_QUERY,
    },
    result: {
      data: {
        me: fakeUser(),
      },
    },
  },
];

describe('<PleaseSignin />', () => {
  it('renders the sign in dialog to logged out users', async () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignedInMocks}>
        <PleaseSignInComponent />
      </MockedProvider>
    );

    expect(wrapper.text()).toContain('Loading...');
    await wait();
    wrapper.update();
    expect(toJSON(wrapper.find('p'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('fieldset'))).toMatchSnapshot();
  });

  it('renders the child component when the user is signed in', async () => {
    const Hey = () => <p>Hey</p>;
    const wrapper = mount(
      <MockedProvider mocks={signedInMocks}>
        <PleaseSignInComponent>
          <Hey />
        </PleaseSignInComponent>
      </MockedProvider>
    );

    await wait();
    wrapper.update();
    expect(wrapper.find('Hey').exists()).toBe(true);
    expect(wrapper.contains(<Hey />)).toBe(true);
  });
});
