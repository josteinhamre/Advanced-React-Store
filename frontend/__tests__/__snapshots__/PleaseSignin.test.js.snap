// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PleaseSignin /> renders the sign in dialog to logged out users 1`] = `
<p>
  Please Sign In before Continuing!
</p>
`;

exports[`<PleaseSignin /> renders the sign in dialog to logged out users 2`] = `
<fieldset
  aria-busy={false}
  disabled={false}
>
  <h2>
    Sign in
  </h2>
  <DisplayError
    error={Object {}}
  />
  <label
    htmlFor="email"
  >
    Email
    <input
      id="email"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      required={true}
      type="email"
      value=""
    />
  </label>
  <label
    htmlFor="password"
  >
    Password
    <input
      id="password"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      required={true}
      type="password"
      value=""
    />
  </label>
  <button
    type="submit"
  >
    Sign In!
  </button>
</fieldset>
`;
